### PR TITLE
ARROW-9782: [C++][Dataset] More configurable Dataset writing

### PR DESCRIPTION
**Pull request by bkietz**
Python:
- ParquetFileFormat.write_options has been removed
- Added classes {,Parquet,Ipc}FileWriteOptions
- FileWriteOptions are constructed using FileFormat.make_write_options(...)
- FileWriteOptions are passed as a parameter to _filesystemdataset_write()

R:
- FileWriteOptions$create(...) to make write options; no subclasses exposed in R
- A filter() on the dataset is applied to restrict written rows.

C++:
- FileSystemDataset::Write's parameters have been consolidated into
  - A Scanner, from which the batches to be written are pulled
  - A FileSystemDatasetWriteOptions, which is an options struct specifying
    - destination filesystem
    - base directory
    - partitioning
    - basenames (via a string template, ex "dat_{i}.feather")
    - format specific write options
- Format specific write options are represented using the FileWriteOptions hierarchy. An instance of these can be constructed from a format using FileFormat::DefaultWriteOptions(), after which the instance can be modified.
  - ParquetFileFormat::{writer_properties, arrow_writer_properties} have been moved to ParquetFileWriteOptions, an implementation of FileWriteOptions.

Internal C++:
- Individual files can now be incrementally written using a FileWriter, constructible from a format using FileFormat::MakeWriter
- FileSystemDataset::Write now parallelizes across scan tasks rather than fragments, so there will be no difference in performance for different arrangements of tables/batches/lists of tables and batches when writing from memory
- FileSystemDataset::Write::WriteQueue provides a threadsafe channel for batches awaiting write, allowing threads to produce batches as another thread flushes the queue to disk.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@bkietz this removes the ability to specify format specific options? (or it's still WIP?)

**Comment by bkietz** [comment]
@jorisvandenbossche yes, not ready for review yet. I will repair format specific write options as part of this PR

**Comment by jorisvandenbossche** [comment]
Okido, will wait a bit more then ;-)

File: cpp/src/arrow/dataset/file_base.cc [hunk 1]
```diff
@@ -143,97 +145,205 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
...
-    }
-  }
+    util::Mutex::Guard writer_lock;
+
+    auto set_lock = mutex_.Lock();
 
-  for (auto&& path_batches : path_to_batches) {
-    auto dir = base_dir + path_batches.first;
-    RETURN_NOT_OK(filesystem->CreateDir(dir, /*recursive=*/true));
+    auto writer =
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/file_base.cc:203]: I don't understand the logic here about dual locking (set_lock, writer_lock).

File: cpp/src/arrow/dataset/file_base.cc [hunk 2]
```diff
@@ -143,97 +145,205 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
...
+  //     an in-memory table to disk.
+  ARROW_ASSIGN_OR_RAISE(FragmentVector fragments, scanner->GetFragments().ToVector());
+  ScanTaskVector scan_tasks;
+  std::vector<const Fragment*> fragment_for_task;
+
+  // Avoid contention with multithreaded readers
+  auto context = std::make_shared<ScanContext>(*scanner->context());
+  context->use_threads = false;
+
+  for (const auto& fragment : fragments) {
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/file_base.cc:276]: Since we know all fragments (and their expressions) already, can we avoid all the locking multi-threading in WriterSet (IIRC, you need them to create the writer once)? That would heavily simplify all of this.

Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc:276]: In this context fragments are the object of writing rather than the target (so for example one might represent an in-memory table which is being copied to disk). Writers are *not* known ahead of time since they depend on the partitioning which depends on the set of unique values in a given column, which we discover only after running GroupBy on an input batch

File: cpp/src/arrow/dataset/file_base.cc [hunk 3]
```diff
@@ -143,97 +145,205 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
   return MakeVectorIterator(std::move(fragments));
 }
 
-struct WriteTask {
-  Status Execute();
+Status FileWriter::Write(RecordBatchReader* batches) {
+  for (std::shared_ptr<RecordBatch> batch;;) {
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc:149]: I personally prefer `for(;;)` to `while(true)` but if it'll be clearer then I'll rewrite it

File: cpp/src/arrow/dataset/file_base.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc:203]: I'll add comments inline

File: cpp/src/arrow/dataset/file_base.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/dataset/file_base.cc:149]: I have a preference for `while (true)` which is much more explicit. Just my 2 cents :-)

File: cpp/src/arrow/dataset/file_base.cc (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc:286]: We *could* do two scans of the input data:
1. Assemble a list of all unique values in the partition columns of the data, from which we can determine the precise set of writers to open
2. Apply groupings to batches, passing the results to pre-opened writers

This doesn't seem worthwhile to me; scanning the input is potentially expensive so we should avoid doing it twice. Furthermore we'll still need to coordinate between threads since two input batches might still contain rows bound for a single output writer.

File: cpp/src/arrow/dataset/file_base.cc [hunk 4]
```diff
@@ -143,97 +150,235 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
...
+    }
+  }
 
-    // make a record batch reader which yields from a fragment
-    ARROW_ASSIGN_OR_RAISE(task->batches, FragmentRecordBatchReader::Make(
-                                             std::move(fragment), schema, scan_context));
-    task_group->Append([task] { return task->Execute(); });
+  // WriteQueues are stored in this deque until writing is completed and are otherwise
+  // referenced by non-owning pointers.
+  std::deque<WriteQueue> queues_storage;
```
Comment (pitrou) [inline cpp/src/arrow/dataset/file_base.cc:299]: This is because `WriteQueue::Set` is an unordered_map, right? If you make it a `map<string, Queue>` instead, Queue pointers won't be invalidated on insert.

File: cpp/src/arrow/dataset/file_base.cc [hunk 5]
```diff
@@ -143,97 +150,235 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
...
+
+            queue = internal::GetOrInsertGenerated(
+                        &queues, std::move(part),
+                        [&](const std::string& emplaced_part) {
+                          // lookup in `queues` also failed,
+                          // generate a new WriteQueue
+                          size_t queue_index = queues.size() - 1;
+
+                          queues_storage.emplace_back(emplaced_part, queue_index,
+                                                      &wait_for_opened_writer_lock);
```
Comment (pitrou) [inline cpp/src/arrow/dataset/file_base.cc:344]: This unorthodox thing with the `wait_for_opened_writer_lock` looks like it's because of two-phase initialization...?

Let me suggest something else: make a `WriteQueuePromise` class:
```c++
class WriteQueuePromise {
 public:
  WriteQueuePromise(std::string part, size_t queue_index,
      std::shared_ptr<Schema> schema, const FileSystemDatasetWriteOptions& options)
    : queue_(part, queue_index), schema_(schema), options_(options) {}

  Result<WriteQueue>* operator()() {
    std::call_once(initialized_, [this]() { InitQueue(); });
    RETURN_NOT_OK(status_);
    return &queue_;
  }

 protected:
  void InitQueue() {
    status_ = queue_->OpenWriter(options_, schema_);
  }

  std::once_flag initialized_;
  WriteQueue queue_;
  std::shared_ptr<Schema> schema_;
  const FileSystemDatasetWriteOptions& options_;
  Status status_;
};
```

Also, make `queues` a `std::map<std::string, WriteQueuePromise>`.
Then do something like:
```c++
WriteQueuePromise* queue_promise;
{
  auto queues_lock = queues_mutex.Lock();
  queue_promise = &queues.emplace(part, part, queue_index, batch->schema(), write_options)->second;
}
queue = (*queue_promise)();
```

Comment (pitrou) [reply cpp/src/arrow/dataset/file_base.cc:344]: It could also be a `DeferredInit` facility:
```c++
template <typename T>
class DeferredInit<T> {
 public:
  template <typename... Args>
  DeferredInit(Args&&... args)
    : value_(std::forward<Args>(args)...) {}

  Result<T*> operator()() {
    std::call_once(initialized_, [this]() { value_.Init(); });
    RETURN_NOT_OK(status_);
    return &value_;
  }

 protected:
  T value_;
  Status status_;
  std::once_flag initialized_;
};
```

File: cpp/src/arrow/dataset/file_base.cc (hunk 4 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc:299]: I was going with unordered_map due to O(1) lookup time; no threads can lookup queues for pushing while I'm holding `queues_lock`

File: cpp/src/arrow/dataset/file_base.cc (hunk 5 above)
Comment (pitrou) [reply cpp/src/arrow/dataset/file_base.cc:344]: If I'm not mistaken, with this approach you don't need the embedded lock guard in `WriteQueue` anymore. Does that make sense?

File: python/pyarrow/dataset.py [hunk 6]
```diff
@@ -694,22 +697,23 @@ def _ensure_write_partitioning(scheme):
     return scheme
 
 
-def write_dataset(data, base_dir, format=None, partitioning=None, schema=None,
-                  filesystem=None, use_threads=True):
+def write_dataset(data, base_dir, basename_template, format=None,
```
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:700]: Should we provide a default template here? 
Can eg the format object have a property with the default name to use? (or get the extension from there and use that in a default?)

File: cpp/src/arrow/dataset/file_base.cc (hunk 5 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc:344]: > Let me suggest something else: make a `WriteQueuePromise` class

That sounds doable and would certainly help with encapsulation of the locks.

> If I'm not mistaken, with this approach you don't need the embedded lock guard in WriteQueue anymore. Does that make sense?

`call_once` does guarantee that side effects of the first invocation are observed by all subsequent invocations, so the FileWriter would indeed be initialized for all threads. However we'd still need a `writer_lock` since `FileWriter::Write` is not thread safe

File: python/pyarrow/dataset.py (hunk 6 above)
Comment (bkietz) [reply python/pyarrow/dataset.py:700]: That sounds fine but I'd like to defer that to a follow up

File: cpp/src/arrow/dataset/file_base.cc (hunk 5 above)
Comment (pitrou) [reply cpp/src/arrow/dataset/file_base.cc:344]: Ah... does the `writer_lock` block inside the thread pool?

Comment (pitrou) [reply cpp/src/arrow/dataset/file_base.cc:344]: Ah, you're using `TryLock`, nevermind :-)

**Review by nealrichardson — APPROVED** [review]
+1 from me, thanks for doing this!

File: cpp/src/arrow/dataset/file_base.cc (hunk 5 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc]: Instead of WriteQueuePromise, I've refactored to initialize the writer in the first call to Flush(), so now all mutexes, locks, and init are internal to WriteQueue.

File: cpp/src/arrow/dataset/file_base.h [hunk 7]
```diff
@@ -128,6 +128,9 @@ class ARROW_DS_EXPORT FileFormat : public std::enable_shared_from_this<FileForma
   /// \brief Return true if fragments of this format can benefit from parallel scanning.
   virtual bool splittable() const { return false; }
 
+  virtual bool Equals(const FileFormat& other) const = 0;
+  //  virtual std::string ToString() const = 0;
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.h:132]: Yes, the theory was to provide a more descriptive error message if ParquetFileWriteOptions were provided to IpcFileFormat::MakeWriter. I'll remove it

File: cpp/src/arrow/dataset/file_ipc.cc [hunk 8]
```diff
@@ -159,18 +163,44 @@ Result<ScanTaskIterator> IpcFileFormat::ScanFile(std::shared_ptr<ScanOptions> op
                                    fragment->source());
 }
 
-Status IpcFileFormat::WriteFragment(RecordBatchReader* batches,
-                                    io::OutputStream* destination) const {
-  ARROW_ASSIGN_OR_RAISE(auto writer, ipc::MakeFileWriter(destination, batches->schema()));
+///
+/// IpcFileWriter, IpcFileWriteOptions
+///
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/file_ipc.cc:168]: Okay

File: cpp/src/arrow/util/string.h [hunk 9]
```diff
@@ -56,5 +57,9 @@ bool AsciiEqualsCaseInsensitive(util::string_view left, util::string_view right)
 ARROW_EXPORT
 std::string AsciiToLower(util::string_view value);
 
+ARROW_EXPORT
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/string.h:60]: Will do

File: cpp/src/arrow/util/mutex.cc [hunk 10]
```diff
@@ -19,25 +19,36 @@
...
 
+#include "arrow/util/logging.h"
+
 namespace arrow {
 namespace util {
 
-struct Mutex::Impl {
-  std::mutex mutex_;
-};
+struct Mutex::Impl : std::mutex {};
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/mutex.cc:27]: I'll revert this

File: cpp/src/arrow/util/mutex.h [hunk 11]
```diff
@@ -31,16 +31,20 @@ namespace util {
...
-  class Guard {
+  class ARROW_EXPORT Guard {
    public:
     Guard() : locked_(NULLPTR, [](Mutex* mutex) {}) {}
     Guard(Guard&&) = default;
     Guard& operator=(Guard&&) = default;
 
     explicit operator bool() const { return bool(locked_); }
 
+    void Unlock();
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/mutex.h:46]: I'd prefer to continue acquiring new locks exclusively through Mutex::Lock; there's no loss of generality and it keeps Guard as simple as possible

**Comment by nealrichardson** [comment]
Is this done, or what is left?

**Comment by bkietz** [comment]
@pitrou are you planning to review C++ again?
@jorisvandenbossche are you planning to review python?

**Comment by pitrou** [comment]
The C++ changes addressed my comments. It would be nice though if @fsaintjacques could take a look.

File: python/pyarrow/_dataset.pyx [hunk 12]
```diff
@@ -2105,68 +2288,41 @@ def _get_partition_keys(Expression partition_expression):
...
-    c_schema = pyarrow_unwrap_schema(schema)
-    c_format = format.unwrap()
-    c_filesystem = filesystem.unwrap()
-    c_partitioning = partitioning.unwrap()
-    c_context = _build_scan_context(use_threads=use_threads)
+    c_options.file_write_options = file_options.unwrap()
+    c_options.filesystem = filesystem.unwrap()
+    c_options.base_dir = tobytes(_stringify_path(base_dir))
+    c_options.partitioning = partitioning.unwrap()
+    c_options.basename_template = tobytes(basename_template)
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:2308]: The `FileFormat` is not part of the c_options? As that argument now seems to be ignored?

Comment (jorisvandenbossche) [reply python/pyarrow/_dataset.pyx:2308]: OK, I suppose the format is embedded in the FileWriteOptions. But then FileFormat should not be passed here, I think? And we should also check that FileFormat and FileWriteOptions don't conflict if both are passed?

File: python/pyarrow/_dataset.pyx [hunk 13]
```diff
@@ -2105,68 +2288,41 @@ def _get_partition_keys(Expression partition_expression):
...
                     c_batches.push_back((<RecordBatch> batch).sp_batch)
             else:
                 c_batches.push_back((<RecordBatch> table).sp_batch)
 
-            c_fragment = shared_ptr[CFragment](
-                new CInMemoryFragment(c_batches, _true.unwrap()))
-            c_batches.clear()
-            c_fragment_vector.push_back(c_fragment)
+        data = Fragment.wrap(shared_ptr[CFragment](
+            new CInMemoryFragment(move(c_batches), _true.unwrap())))
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:2322]: Does this change behaviour? It seems you are now creating a single fragment instead of a vector of fragments?

File: python/pyarrow/dataset.py [hunk 14]
```diff
@@ -741,6 +746,12 @@ def write_dataset(data, base_dir, format=None, partitioning=None, schema=None,
         )
 
     format = _ensure_format(format)
+    if basename_template is None:
+        basename_template = "dat_{i}." + format.default_extname
```
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:750]: Maybe we could go with `"part-{i}."` for the default name (that seems more in line with eg dask / spark) 
(I can also do this in a follow-up PR, if we want this)

**Review by jorisvandenbossche — COMMENTED** [review]
Did a pass over the python code

File: python/pyarrow/_dataset.pyx (hunk 12 above)
Comment (bkietz) [reply python/pyarrow/_dataset.pyx:2308]: I'll remove the unused parameters and a check for format consistency in write_dataset

File: python/pyarrow/_dataset.pyx [hunk 15]
```diff
@@ -1054,18 +1095,173 @@ cdef class ParquetReadOptions(_Weakrefable):
...
+        arrow_fields = {
+            "use_deprecated_int96_timestamps",
+            "coerce_timestamps",
+            "allow_truncated_timestamps",
+        }
+
+        update = False
+        update_arrow = False
+        for name, value in kwargs.items():
+            assert name in self._properties
```
[the remark this answers is not in the document]
Comment (bkietz) [reply python/pyarrow/_dataset.pyx:1116]: will do

File: python/pyarrow/_dataset.pyx [hunk 16]
```diff
@@ -1054,18 +1095,173 @@ cdef class ParquetReadOptions(_Weakrefable):
...
+                writer_engine_version=self.writer_engine_version,
+            )
+
+    @property
+    def use_dictionary(self):
+        return self._properties['use_dictionary']
+
+    @use_dictionary.setter
+    def use_dictionary(self, use_dictionary):
+        self.update(use_dictionary=use_dictionary)
```
[the remark this answers is not in the document]
Comment (bkietz) [reply python/pyarrow/_dataset.pyx:1151]: I'll remove the getters and setters.

File: python/pyarrow/_dataset.pyx (hunk 13 above)
Comment (bkietz) [reply python/pyarrow/_dataset.pyx:2322]: FileSystemDataset::Write now parallelizes across scan tasks rather than fragments so there will be no difference in performance/written files even if we create a single in-memory fragment, so I changed this to create a single fragment since it's simpler

File: python/pyarrow/dataset.py (hunk 14 above)
Comment (bkietz) [reply python/pyarrow/dataset.py:750]: Not a problem, I'll push this shortly

**Comment by bkietz** [comment]
Merging

File: python/pyarrow/tests/test_dataset.py [hunk 17]
```diff
@@ -2200,7 +2196,7 @@ def test_write_dataset(tempdir):
     dataset = ds.dataset(directory)
 
     target = tempdir / 'single-directory-target'
-    expected_files = [target / "dat_0.ipc", target / "dat_1.ipc"]
+    expected_files = [target / "part-0.feather"]
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:2199]: Why did this change to a single file? (the original has 2 files, I expect the roundtrip to preserve those files)

File: python/pyarrow/tests/test_dataset.py [hunk 18]
```diff
@@ -2258,11 +2254,6 @@ def test_write_dataset_use_threads(tempdir):
         use_threads=False
     )
 
-    # check that all files are the same in both cases
-    paths1 = [p.relative_to(target1) for p in target1.rglob("*")]
-    paths2 = [p.relative_to(target2) for p in target2.rglob("*")]
-    assert set(paths1) == set(paths2)
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:2264]: Why was this removed? (does it no longer hold?)

**Comment by jorisvandenbossche** [comment]
> FileSystemDataset::Write now parallelizes across scan tasks rather than fragments so there will be no difference in performance/written files even if we create a single in-memory fragment, so I changed this to create a single fragment since it's simpler

There might be no difference, but I think the user should still be able to control how many files are created. Because now whathever you pass, it's always consolidated into a single file? (or one file per partitioning) 

Also, it seems that reading/writing a dataset does not preserve the files? (so if we discover a dataset with multiple files, we write it as a single file?)

File: python/pyarrow/tests/test_dataset.py (hunk 17 above)
Comment (bkietz) [reply python/pyarrow/tests/test_dataset.py:2199]: After this patch, a single file will be written for each partition directory. In a follow up we'll add an optional cap on file size

File: python/pyarrow/tests/test_dataset.py (hunk 18 above)
Comment (bkietz) [reply python/pyarrow/tests/test_dataset.py:2264]: It no longer holds consistently; the auto incremented `{i}` doesn't necessarily round trip

**Comment by bkietz** [comment]
If you're writing with no partitioning then yes, everything will be written to a single file. In a follow up we'll probably add a special case for unpartitioned writing which allocates an output file for each thread just for performance reasons.
